File: packages/host4_flutter_aivoice/ios/AiVoice_副本/View/AiViewManager/AiViewManager.m
//
//  AiViewManager.m
//  GameMacro
//
//  Created by 敬攀 on 2025/12/16.
//

#import "AiViewManager.h"
#import "XHFloatWindow.h"
#import "XHFloatWindowController.h"
#import "AiVoiceManager.h"
#import "FrameAnimationView.h"
#import "FrameAnimationView+Animations.h"
#import "ChatAssistantView.h"
#import "UserSignInVC.h"


@interface AiViewManager()<ChatAssistantViewDelegate>

@end

static AiViewManager *viewManager = nil;

@implementation AiViewManager

+(AiViewManager *)shared{
    static dispatch_once_t onceToken;
    
    dispatch_once(&onceToken, ^{
        viewManager = [[AiViewManager alloc] init];
        
        //初始状态都是FloatingState
        viewManager.floatState = FloatingState;
    });
    
    return viewManager;
}

#pragma mark - 悬浮球
-(void)showFloatWindow:(BOOL)isshow{
    
    NSLog(@"显示悬浮球 floatState: %ld", (long)self.floatState);
    
    BaseVC *target = (BaseVC *)[TheGlobalMethod getCurrentVC];
    
    if (!isshow){
        //不显示的时候直接销毁
        [XHFloatWindow xh_destroyWindow];

        if ([ChatAssistantView shared].isShowing){
            [[ChatAssistantView shared] hide];
        }
        
        return;
    }
    
    if (self.floatState != FloatingState){
        //聊天状态,显示聊天
        [self showChatFloatCustomView];
    }else{
        //悬浮窗状态
        kWeakSelf(self)
        
        if ([XHFloatWindow xh_isShowing]){
            return;
        }
        
        [XHFloatWindow xh_addWindowOnTarget:target onClick:^{
            
            if ([weakself isPushNothing]){
                [MBProgressHUD showMsg:kLocalizedString(@"LoginAndUse", nil) toView:KEY_WINDOW];
            }else{
                CGRect frame = [XHFloatWindow xh_getCurrentFrame];
                
                [XHFloatWindow xh_destroyWindow];

                [self playAnimalImage:frame];
            }
        }];
        
        //根据缩小放大状态
        BOOL isShrunk = [XHFloatWindow xh_getSavedShrunkState];
        
        CGFloat floatWidth;
        
        if (isShrunk){
            //缩小状态
            floatWidth = kShrunkSize;
        }else{
            //放大状态
            floatWidth = kNormalSize;
        }
        
        [XHFloatWindow xh_setWindowSize:floatWidth];
        
        [XHFloatWindow xh_setBackgroundImage:@"AINotClick" forState:(UIControlStateNormal)];
        
        if (SCREEN_H > SCREEN_W){
            //竖屏(默认显示横屏放大状态)
            [XHFloatWindow xh_setInitialPosition:CGPointMake((SCREEN_MIN - floatWidth) / 2, SCREEN_MAX - 100 * COEFi0 - floatWidth / 2)];
        }else{
            if (isShrunk){
                //横屏放大
                [XHFloatWindow xh_setInitialPosition:CGPointMake(SCREEN_MAX - floatWidth, 65 * COEFi0)];
            }
        }
        
    }
}

//帧动画视图
- (void)playAnimalImage:(CGRect)frame{
    
    //如果没有登陆，那么跳转登陆
    if (![TheGlobalMethod getModelInLocal:GMUserModel]){
        //跳转登陆
        [self pushController:0];
        
        return;
    }
    
    FrameAnimationView *view = [[FrameAnimationView alloc] initWithImagePrefix:@"思考状态-出现_" startIndex:0 count:124 numberFormat:@"%05d" extension:@"png"];
    
    view.frameRate = 40;
    
    view.animationMode = FrameAnimationModeOnce;
    
    [KEY_WINDOW addSubview:view];
      
    CGPoint center = CGPointMake(frame.origin.x + frame.size.width / 2, frame.origin.y + frame.size.height / 2);
    
    [view animateFromCenter:center toSize:CGSizeMake(100 * COEFi0, 100 * COEFi0) duration:3 damping:3 autoPlay:YES completion:^{
        //动画结束，销毁视图，并创建聊天视图
        [self showFloatWindow:NO];
        
        [view removeFromSuperview];
        
        self.floatState = ChatMinimized;
        
        [self showChatFloatCustomView];
    }];
}

//聊天界面
-(void)showChatFloatCustomView{
    [AIVoiceShared buildRTCEngine];
    
    [ChatAssistantView shared].delegate = self;
    
    switch (self.floatState) {
        case ChatMinimized:
            [[ChatAssistantView shared] showMinimizedAtScreen];
            break;
            
        case ChatExpanded:
            [[ChatAssistantView shared] showExpandedAtScreen];
            break;
            
        default:
            break;
    }
}

-(void)setFloatState:(AiFloatingViewState)floatState{
    _floatState = floatState;
}

#pragma mark - ChatAssistantViewDelegate

//关闭按钮
-(void)chatAssistantViewCloseButtonTapped:(ChatAssistantView *)view{
    
    [self showFloatWindow:NO];
    
    [ChatAssistantView destroy];
    
    [AIVoiceShared destructionRTCEngine];
    
    [AiViewManager shared].floatState = FloatingState;
    
    QueueStartAfterTime(.5)
    
    [AIVoiceShared destructionRTCEngine];
    
    [self showFloatWindow:YES];
    
    queueEnd
}

//音量按钮
-(void)chatAssistantViewVolumeButtonTapped:(ChatAssistantView *)view isSec:(BOOL)isSec{

    [AIVoiceShared switchVoiceVolume:!isSec];
}

- (void)chatAssistantViewVipButtonTapped:(ChatAssistantView *)view{
    //跳转vip
    [self pushController:1];
}

//-(void)pushController:(int)type{
////    BOOL isDelay = SCREEN_W > SCREEN_H;
////    
////    if (isDelay){
////        [[LandscapeStatusView shareView] hiddenView];
////        
////        [[LandscapeToolBar shareToolBar] hiddenView];
////    }
//
//    BaseVC *currentVC = (BaseVC *)[TheGlobalMethod getCurrentController];
//    
//    if (type == 0){
//        //登陆界面
////        UserSignInVC *signVC = [[UserSignInVC alloc] initWithDelay:isDelay];
//        UserSignInVC *signVC = [[UserSignInVC alloc] init];
//        
//        [currentVC.navigationController pushViewController:signVC animated:YES];
//        
////        signVC.popController = ^{
////            [self showFloatWindow:NO];
////            
////            QueueStartAfterTime(.5)
////            
////            [self showFloatWindow:YES];
////            
////            queueEnd
////        };
//    }
//    
//    if (type == 1){
//        //vip界面
//        
//        if (!isStringEmpty([HttpRequrst shareManager].appM.H5_VIP)){
//            
////            WKViewController *webVC = [[WKViewController alloc] initWithUrl:[HttpRequrst shareManager].appM.H5_VIP isShowNav:NO isDelay:isDelay];
//            WKViewController *webVC = [[WKViewController alloc] initWithUrl:[HttpRequrst shareManager].appM.H5_VIP isShowNav:NO isDelay:NO];
//            
//            [currentVC.navigationController pushViewController:webVC animated:YES];
//            
////            webVC.popController = ^{
////                [self showFloatWindow:NO];
////                
////                currentVC.isLaunchScreen = YES;
////                
////                QueueStartAfterTime(.5)
////                
////                [self showFloatWindow:YES];
////                
////                queueEnd
////            };
//        }
//    }
//    
//    [self showFloatWindow:NO];
//    
//    [ChatAssistantView destroy];
//    
//    [AIVoiceShared destructionRTCEngine];
//    
//    [AiViewManager shared].floatState = FloatingState;
//    
////    CGFloat delayTime = isDelay ? 1 : 0;
//    
////    QueueStartAfterTime(delayTime)
//    
//    [self showFloatWindow:YES];
//    
////    queueEnd
//    
//}

-(void)pushController:(int)type{
    UINavigationController *nav = [TheGlobalMethod getCurrentNavController];
    
    if (!nav) {
        [MBProgressHUD showMsg:@"导航控制器未就绪"];
        return;
    }

    // 获取当前显示的VC
    UIViewController *topVC = nav.topViewController;
    
    dispatch_async(dispatch_get_main_queue(), ^{
        
        [self showFloatWindow:YES];
        
        if (type == 0){
            // 已经是登录页就不再 push
            if ([topVC isKindOfClass:[UserSignInVC class]]) {
                return;
            }
            
            UserSignInVC *signVC = [[UserSignInVC alloc] init];
            [nav pushViewController:signVC animated:YES];
        }
        if (type == 1){
            if (!isStringEmpty([HttpRequrst shareManager].appM.H5_VIP)){
                // 已经是 VIP 页就不再 push
                if ([topVC isKindOfClass:[WKViewController class]]) {
                    return;
                }
                
                WKViewController *webVC = [[WKViewController alloc] initWithUrl:[HttpRequrst shareManager].appM.H5_VIP isShowNav:NO isDelay:NO];
                [nav pushViewController:webVC animated:YES];
            }
        }
        
//        [self showFloatWindow:NO];
//        [ChatAssistantView destroy];
//        [AIVoiceShared destructionRTCEngine];
//        [AiViewManager shared].floatState = FloatingState;
//        [self showFloatWindow:YES];
        
    });
}

-(BOOL)isPushNothing{
    BaseVC *currentVC = (BaseVC *)[TheGlobalMethod getCurrentController];
    
    if (![TheGlobalMethod getModelInLocal:GMUserModel] && [currentVC isKindOfClass:[UserSignInVC class]]){
        //跳转登陆
        return YES;
    }
    
    return NO;
}

@end

File: packages/host4_flutter_aivoice/ios/Classes/Views/AiViewManager.swift
import UIKit

/// AI 悬浮窗/聊天窗状态
enum AiFloatingViewState {
  case floating    // 悬浮球
  case chatMinimized  // 聊天窗缩小
  case chatExpanded   // 聊天窗展开
}

/// AI 视图管理器（单例）
/// 管理悬浮球和聊天窗的显示/隐藏/状态切换
class AiViewManager: NSObject {

  static let shared = AiViewManager()

  /// 当前状态
  var floatState: AiFloatingViewState = .floating

  /// 兼容 iOS 13+ 的 keyWindow
  private var currentKeyWindow: UIWindow? {
    if #available(iOS 13.0, *) {
      return UIApplication.shared.connectedScenes
        .compactMap { $0 as? UIWindowScene }
        .flatMap { $0.windows }
        .first { $0.isKeyWindow }
    } else {
      return UIApplication.shared.keyWindow
    }
  }

  private override init() {
    super.init()
  }

  /// 显示/隐藏 AI 悬浮窗
  /// - Parameter show: true=显示, false=销毁
  func showFloatWindow(_ show: Bool) {
    print("[AiView] showFloatWindow: \(show), state: \(floatState)")

    let targetVC = UIApplication.shared.topViewController(base: nil)

    if !show {
      XHFloatWindow.xh_destroyWindow()
      if ChatAssistantView.shared.isShowing {
        ChatAssistantView.shared.hide()
      }
      AiVoiceManager.shared.destructionRTCEngine()
      return
    }

    switch floatState {
    case .floating:
      showFloatingBall(target: targetVC)
    case .chatMinimized, .chatExpanded:
      showChatView()
    }
  }

  /// 显示悬浮球
  private func showFloatingBall(target: UIViewController?) {
    if XHFloatWindow.xh_isShowing() { return }

    XHFloatWindow.xh_addWindowOnTarget(target ?? UIViewController()) { [weak self] in
      guard let self = self else { return }
      if self.isPushNothing() {
        self.pushController(0)
      } else {
        let frame = XHFloatWindow.xh_getCurrentFrame()
        XHFloatWindow.xh_destroyWindow()
        self.playAnimation(from: frame)
      }
    }

    // 根据历史状态设置大小
    let isShrunk = XHFloatWindow.xh_getSavedShrunkState()
    let floatWidth = isShrunk ? kShrunkSize : kNormalSize
    XHFloatWindow.xh_setWindowSize(floatWidth)
    XHFloatWindow.xh_setBackgroundImage("AINotClick", for: .normal)

    let screenW = UIScreen.main.bounds.width
    let screenH = UIScreen.main.bounds.height

    if screenH > screenW {
      // 竖屏
      let minW = min(screenW, screenH)
      XHFloatWindow.xh_setInitialPosition(CGPoint(
        x: (minW - floatWidth) / 2,
        y: screenH - 100 * kCoefi - floatWidth / 2
      ))
    } else {
      if isShrunk {
        XHFloatWindow.xh_setInitialPosition(CGPoint(
          x: max(screenW, screenH) - floatWidth,
          y: 65 * kCoefi
        ))
      }
    }
  }

  /// 播放帧动画（点击悬浮球后的展开动画）
  private func playAnimation(from frame: CGRect) {
    DispatchQueue.main.async { [weak self] in
      guard let self = self else { return }
      // 帧动画: 使用 "思考状态-出现_" + 124 帧 png
      let animView = FrameAnimationView(
        imagePrefix: "思考状态-出现_",
        startIndex: 0,
        count: 124,
        numberFormat: "%05d",
        extension: "png"
      )
      animView.frameRate = 40
      animView.animationMode = .once

      guard let window = self.currentKeyWindow else { return }
      window.addSubview(animView)

      let center = CGPoint(
        x: frame.origin.x + frame.size.width / 2,
        y: frame.origin.y + frame.size.height / 2
      )

      animView.animate(
        fromCenter: center,
        toSize: CGSize(width: 100 * kCoefi, height: 100 * kCoefi),
        duration: 3,
        damping: 3,
        autoPlay: true,
        completion: { [weak self] in
          self?.showFloatWindow(false)
          animView.removeFromSuperview()
          self?.floatState = .chatMinimized
          self?.showChatView()
        }
      )
    }
  }

  /// 显示聊天界面
  func showChatView() {
    AiVoiceManager.shared.buildRTCEngine()
    ChatAssistantView.shared.delegate = self

    switch floatState {
    case .chatMinimized:
      ChatAssistantView.shared.showMinimizedAtScreen()
    case .chatExpanded:
      ChatAssistantView.shared.showExpandedAtScreen()
    default:
      break
    }
  }

  /// 需要登录/VIP 页面时触发的回调（由外部宿主 App 处理页面跳转）
  var onNeedPushLogin: (() -> Void)?
  var onNeedPushVip: (() -> Void)?

  /// 跳转页面
  func pushController(_ type: Int) {
    DispatchQueue.main.async { [weak self] in
      guard let self = self else { return }
      self.showFloatWindow(true)

      if type == 0 {
        self.onNeedPushLogin?()
      } else if type == 1 {
        self.onNeedPushVip?()
      }
    }
  }

  /// 是否需要拦截点击跳转（如未登录状态）
  /// 外部可通过此回调返回是否拦截
  var onCheckShouldBlockTap: (() -> Bool)?

  func isPushNothing() -> Bool {
    return onCheckShouldBlockTap?() ?? false
  }
}

// MARK: - ChatAssistantViewDelegate
extension AiViewManager: ChatAssistantViewDelegate {
  func chatAssistantViewCloseButtonTapped(_ view: ChatAssistantView) {
    showFloatWindow(false)
    ChatAssistantView.destroy()
    AiVoiceManager.shared.destructionRTCEngine()
    floatState = .floating

    DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
      self?.showFloatWindow(true)
    }
  }

  func chatAssistantViewVolumeButtonTapped(_ view: ChatAssistantView, isSec: Bool) {
    AiVoiceManager.shared.switchVoiceVolume(!isSec)
  }

  func chatAssistantViewFullscreenButtonTapped(_ view: ChatAssistantView) {
    // 切换展开/缩小
  }

  func chatAssistantViewVipButtonTapped(_ view: ChatAssistantView) {
    pushController(1)
  }
}

// MARK: - Constants
let kShrunkSize: CGFloat = 60
let kNormalSize: CGFloat = 80
let kCoefi: CGFloat = UIScreen.main.bounds.width / 375.0
